### ISSUE TITLE
Update logits array in-place

### ISSUE THREAD

**Issue by brandonwillard** [dottxt-ai/outlines#859]
### What behavior of the library made you think about the improvement?

The current structured generation code is creating a `-inf` copy of the logits array and setting the allowed token ID indices to the corresponding values in the original logits array.  See [here](https://github.com/outlines-dev/outlines/blob/51064f3dd4082a48399150ec9c78ac219116db51/outlines/generate/generator.py#L305).

### How would you like it to behave?

When possible, the original logits array should be updated in-place and completely avoid creating a new array.  This change would likely require the set of disallowed token IDs instead of the allowed ones.

**Comment by brandonwillard** [comment]
@rlouf @lapp0

**Comment by lapp0** [comment]
This is dead code which should be cleaned up after

https://github.com/dottxt-ai/outlines/pull/1010

**Comment by brandonwillard** [comment]
> This is dead code which should be cleaned up after
> 
> #1010

The relevant code now seems to be [here](https://github.com/dottxt-ai/outlines/blob/6035e86ac8089d4f8aeab07ea116093a2ed0e03e/outlines/processors/structured.py#L105); is that what will be updated in #1010?

**Comment by lapp0** [comment]
No, that is a problem, thanks for pointing it out. I'll look at it once my current task is complete.

**Comment by brandonwillard** [comment]
#1192 does not appear to update the logits arrays in place.

**Comment by brandonwillard** [comment]
Looks like this might need to be moved after https://github.com/dottxt-ai/outlines/issues/1206.
